Checkout/Checkout/Source/Model/TokenDetails.swift

```diff
@@ -25,6 +25,25 @@ public struct TokenDetails: Equatable {
   public let billingAddress: Address?
   public let phone: Phone?
   public let name: String?
+    
+  public init(type: TokenType, token: String, expiresOn: String, expiryDate: ExpiryDate, scheme: Card.Scheme?, last4: String, bin: String, cardType: String?, cardCategory: String?, issuer: String?, issuerCountry: String?, productId: String?, productType: String?, billingAddress: Address?, phone: Phone?, name: String?) {
```

Comment:
Hi @Aecasorg , thank you for your involvement and attempt at making the SDK work better for our users. 

I see you explained the PR as 
> Add public init to TokenDetails in order to facilitate access to struct from outside package. In this instance it would be used to aid mocking test results in main project where Frames package is used.

We have previously touched on this internally and the issue is we are not comfortable with introducing the ability for consumers to create mock tokens outside of testing environments. This would make it too easy for implementations that reach production environments to create false outcomes, violating the reliability of the data our SDK outputs.

For unit testing you may be able to use specific Swift syntax to improve access control in a way that facilitates your requirements.

Reply:
Thank you for your response! 

I completely get your concern regarding mocking token outside testing environments. 

I've extended the `CheckoutAPIService` with a protocol in order to access the `createCardToken(card: Card)` function for mocking. Before Frames 4.0 I could to create a `CkoCardTokenResponse` from json, but that is less easy with `TokenDetails`. Did you have any "specific Swift syntax" in mind that would aid with this?

Any help with the above would be most appreciated.

Reply:
Hi @Aecasorg by "specific Swift syntax" I was thinking of `@testable import Frames` inside your unit test target. Having tried it now myself I found a side effect of our type alias definition for `TokenDetails`, which means you need to import a different module to access the initialiser.

I have been trying to see how we can help and ended up with this helper object and sample test to validate it compiles and builds:

## Helper Object inside your unit test module
`TokenCreator.swift`

```Swift
@testable import Checkout

enum TokenCreator {

    static func create(type: TokenDetails.TokenType,
                       token: String = "",
                       expiresOn: String = "",
                       expiryDate: ExpiryDate = ExpiryDate(month: 12, year: 2055),
                       scheme: Card.Scheme? = nil,
                       last4: String = "",
                       bin: String = "",
                       cardType: String? = nil,
                       cardCategory: String? = nil,
                       issuer: String? = nil,
                       issuerCountry: String? = nil,
                       productId: String? = nil,
                       productType: String? = nil,
                       billingAddress: Address? = nil,
                       phone: TokenDetails.Phone? = nil,
                       name: String? = nil) -> TokenDetails {
        TokenDetails(type: type,
                     token: token,
                     expiresOn: expiresOn,
                     expiryDate: expiryDate,
                     scheme: scheme,
                     last4: last4,
                     bin: bin,
                     cardType: cardType,
                     cardCategory: cardCategory,
                     issuer: issuer,
                     issuerCountry: issuerCountry,
                     productId: productId,
                     productType: productType,
                     billingAddress: billingAddress,
                     phone: phone,
                     name: name)
    }

}
```

## Sample unit test to validate it all compiles & works
```Swift
func testExample() {
    let testName = "testToken"
    let token = TokenCreator.create(type: .applePay, name: testName)
    XCTAssertEqual(token.name, testName)
}
```

Please let me know if this works for your requirements. I believe this should enable you to mock data for unit testing, whilst preserving the integrity of SDK output.

Do be warned in a future version we may be dropping the module `Checkout` altogether (nothing confirmed or planned right now), which may require to import Frames instead, but we would still aim to enable integration unit tests that rely on our SDKs output.

Reply:
Thanks for this! This looks helpful. Unfortunately I still get the "Module 'Checkout' was not compiled for testing" error so I can't test it. It looks like an enable testing flag is needed to be added to the SPM Package.swift file in order to enable any external testable access to the Checkout package. 

```
.target(
            name: "Checkout",
            dependencies: [
                "CheckoutEventLoggerKit",
            ],
            path: "Checkout/Checkout/Source",
            swiftSettings: [
                .unsafeFlags(["-enable-testing"])
            ]
        )
```

I have not been able to test and verify this as Xcode refuses to access my GitHub account for adding an SPM. Is there any way this could be tested?